crates/starknet-devnet-server/src/api/json_rpc/endpoints_ws.rs

```diff
@@ -0,0 +1,111 @@
+use starknet_core::error::Error;
+use starknet_rs_core::types::{BlockId, BlockTag};
+use starknet_types::starknet_api::block::BlockStatus;
+
+use super::error::ApiError;
+use super::models::{BlockIdInput, SubscriptionIdInput};
+use super::{JsonRpcHandler, JsonRpcSubscriptionRequest};
+use crate::rpc_core::request::Id;
+use crate::subscribe::{SocketId, SubscriptionNotification};
+
+/// The definitions of JSON-RPC read endpoints defined in starknet_ws_api.json
+impl JsonRpcHandler {
+    pub async fn execute_ws(
+        &self,
+        request: JsonRpcSubscriptionRequest,
+        rpc_request_id: Id,
+        socket_id: SocketId,
+    ) -> Result<(), ApiError> {
+        match request {
+            JsonRpcSubscriptionRequest::NewHeads(data) => {
+                self.subscribe_new_heads(data, rpc_request_id, socket_id).await
+            }
+            JsonRpcSubscriptionRequest::TransactionStatus => todo!(),
+            JsonRpcSubscriptionRequest::PendingTransactions => todo!(),
+            JsonRpcSubscriptionRequest::Events => todo!(),
+            JsonRpcSubscriptionRequest::Unsubscribe(SubscriptionIdInput { subscription_id }) => {
+                let mut sockets = self.api.sockets.lock().await;
+                let socket_context = sockets.get_mut(&socket_id).ok_or(
+                    ApiError::StarknetDevnetError(Error::UnexpectedInternalError {
+                        msg: format!("Unregistered socket ID: {socket_id}"),
+                    }),
+                )?;
+
+                socket_context.unsubscribe(rpc_request_id, subscription_id).await?;
+                Ok(())
+            }
+        }
+    }
+
+    /// starknet_subscribeNewHeads
+    /// Checks if an optional block ID is provided. Validates that the block exists and is not too
+    /// many blocks in the past. If it is a valid block, the user is notified of all blocks from the
+    /// old up to the latest, and subscribed to new ones. If no block ID specified, the user is just
+    /// subscribed to new blocks.
+    pub async fn subscribe_new_heads(
+        &self,
+        block_id_input: Option<BlockIdInput>,
+        rpc_request_id: Id,
+        socket_id: SocketId,
+    ) -> Result<(), ApiError> {
+        let latest_tag = BlockId::Tag(BlockTag::Latest);
+        let block_id = if let Some(BlockIdInput { block_id }) = block_id_input {
+            block_id.into()
+        } else {
+            // if no block ID input, this eventually just subscribes the user to new blocks
+            latest_tag
+        };
+
+        let starknet = self.api.starknet.lock().await;
+
+        // checking the block's existence; aborted blocks treated as not found
+        let query_block = match starknet.get_block(&block_id) {
+            Ok(block) => match block.status() {
+                BlockStatus::Rejected => Err(ApiError::BlockNotFound),
+                _ => Ok(block),
+            },
+            Err(Error::NoBlock) => Err(ApiError::BlockNotFound),
```

Comment:
what happens if block_id is some future block like 2 blocks after the latest.

Reply:
Asked Alex, he [replied last week](https://spaceshard.slack.com/archives/C03QN20522D/p1730113533797169?thread_ts=1729779693.200149&cid=C03QN20522D). It's undocumented behavior currently, but from how I understood his response, a new error (invalid block id) will be added to the spec. So the question is whether we already handle this case or when the spec is updated.